src/libexpr/attr-set.cc

```diff
@@ -2,51 +2,83 @@
 #include "eval-inline.hh"
 
 #include <algorithm>
-
+#include <string.h>
 
 namespace nix {
 
+std::unordered_set<Bindings::Names, Bindings::NamesHasher>       * Bindings::pnamesTable = new std::unordered_set<Bindings::Names, Bindings::NamesHasher>      (0x4000, Bindings::NamesHasher());
+std::unordered_set<std::vector<const Pos*>, Bindings::PosHasher> * Bindings::pposTable   = new std::unordered_set<std::vector<const Pos*>, Bindings::PosHasher>(0x4000, Bindings::PosHasher());
 
-/* Allocate a new array of attributes for an attribute set with a specific
-   capacity. The space is implicitly reserved after the Bindings
-   structure. */
-Bindings * EvalState::allocBindings(size_t capacity)
-{
-    if (capacity > std::numeric_limits<Bindings::size_t>::max())
-        throw Error("attribute set of size %d is too big", capacity);
-    return new (allocBytes(sizeof(Bindings) + sizeof(Attr) * capacity)) Bindings((Bindings::size_t) capacity);
-}
+int Bindings::nNameLists = 0;
+int Bindings::bNameLists = 0;
+int Bindings::nPosLists = 0;
+int Bindings::bPosLists = 0;
 
 
-void EvalState::mkAttrs(Value & v, size_t capacity)
-{
-    if (capacity == 0) {
-        v = vEmptySet;
-        return;
+Bindings* BindingsBuilder::resultAt(void *p, bool alreadySorted) {
+    size_t savedsize = size();
+
+    if (!alreadySorted)
+      quicksort(begin(), end());
+
+#ifndef _NDEBUG
+    resultCalled = true; // size(), begin(), end() do not work after `names` moved to Bindings::pnamesTable
+#endif
+
+    std::pair<std::unordered_set<Bindings::Names, Bindings::NamesHasher>::iterator, bool> res1 = Bindings::pnamesTable->emplace(std::move(names));
+    if (res1.second) {
+        Bindings::nNameLists ++;
+        Bindings::bNameLists += savedsize * sizeof(Symbol);
     }
-    clearValue(v);
-    v.type = tAttrs;
-    v.attrs = allocBindings(capacity);
-    nrAttrsets++;
-    nrAttrsInAttrsets += capacity;
+
+    std::pair<std::unordered_set<std::vector<const Pos*>, Bindings::PosHasher>::iterator, bool> res2 = Bindings::pposTable->emplace(std::move(positions));
+    if (res2.second) {
+        Bindings::nPosLists ++;
+        Bindings::bPosLists += savedsize * sizeof(Pos*);
+    }
+
+    Bindings * b = new (p) Bindings(&*res1.first, &*res2.first);
+    memcpy(&b->values[0], &values[0], savedsize * sizeof(Value *));
+    return b;
 }
 
+Bindings * BindingsBuilder::result(bool alreadySorted) {
+    return resultAt(allocBytes(sizeof(Bindings) + size()*sizeof(Value*)), alreadySorted);
+}
 
-/* Create a new attribute named 'name' on an existing attribute set stored
-   in 'vAttrs' and return the newly allocated Value which is associated with
-   this attribute. */
-Value * EvalState::allocAttr(Value & vAttrs, const Symbol & name)
-{
-    Value * v = allocValue();
-    vAttrs.attrs->push_back(Attr(name, v));
-    return v;
+// own sort function because std::sort has no custom `swap' function :(
+void BindingsBuilder::quicksort(iterator a, iterator b) {
```

Comment:
wow. Stock `std::sort` should be more efficient, so I'm surprised there is performance gain in the end.

Reply:
Copy-pasting code from ```std::sort``` could improve performance even more.